### PR TITLE
ios: advanced server configuration

### PR DESCRIPTION
**Pull request by epoberezkin**

### CODE REVIEW

File: apps/ios/Shared/Views/UserSettings/SMPServerView.swift [hunk 1]
```diff
@@ -24,47 +56,61 @@ struct SMPServerView: View {
...
         }
     }
 
-    private func useServerSection() -> some View {
+    private func useServerSection(_ valid: Bool) -> some View {
         Section("Use server") {
             HStack {
                 Button("Test server") {
-                    Task { await testServerConnection(server: $server) }
+                    Task {
```
Comment (jr-simplex) [inline apps/ios/Shared/Views/UserSettings/SMPServerView.swift:102]: in this view progress indicator is not displayed when testing a previously tested server

File: apps/ios/Shared/Views/UserSettings/SMPServerView.swift [hunk 2]
```diff
@@ -24,47 +56,61 @@ struct SMPServerView: View {
...
+                            Spacer()
+                            Image(systemName: "exclamationmark.circle").foregroundColor(.red)
+                        }
+                    }
                 }
-                useServerSection()
+                useServerSection(valid)
                 Section("Add to another device") {
-                    QRCode(uri: server.server)
+                    MutableQRCode(uri: $serverToEdit.server)
```
Comment (jr-simplex) [inline apps/ios/Shared/Views/UserSettings/SMPServerView.swift:91]: QR code is displayed even on empty string - maybe it's better to display this section only when server can be parsed

File: apps/ios/Shared/Views/UserSettings/SMPServerView.swift (hunk 1 above)
Comment (jr-simplex) [reply apps/ios/Shared/Views/UserSettings/SMPServerView.swift:102]: actually on a new server as well

File: apps/ios/Shared/Views/UserSettings/SMPServerView.swift [hunk 3]
```diff
@@ -10,9 +10,41 @@ import SwiftUI
...
+    @State var testFailure: SMPTestFailure?
 
     var body: some View {
+        smpServerView()
+            .navigationBarBackButtonHidden(true)
+            .toolbar {
+                ToolbarItem(placement: .navigationBarLeading) {
+                    Button {
+                        if serverToEdit.server != server.server {
+                            print("serverToEdit.tested = nil")
```
Comment (jr-simplex) [inline apps/ios/Shared/Views/UserSettings/SMPServerView.swift:26]:
```suggestion
```

File: apps/ios/Shared/Views/UserSettings/SMPServerView.swift [hunk 4]
```diff
@@ -10,9 +10,41 @@ import SwiftUI
...
+                            serverToEdit.tested = nil
```
Comment (jr-simplex) [inline apps/ios/Shared/Views/UserSettings/SMPServerView.swift:27]: editing server doesn't update `showTestStatus` in the view

File: apps/ios/Shared/Views/UserSettings/ScanSMPServer.swift [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+            Text("Scan server QR code")
+                .font(.largeTitle)
+                .bold()
+                .padding(.vertical)
+            ZStack {
+                CodeScannerView(codeTypes: [.qr], completion: processQRCode)
+                    .aspectRatio(1, contentMode: .fit)
+                    .border(.gray)
+            }
+        }
```
Comment (jr-simplex) [inline apps/ios/Shared/Views/UserSettings/ScanSMPServer.swift:29]: Maybe add some text below scanner similar to Connect views? The view feels unfinished atm. For example
```suggestion
        VStack(alignment: .leading) {
            Text("Scan server QR code")
                .font(.largeTitle)
                .bold()
                .padding(.vertical)
            ZStack {
                CodeScannerView(codeTypes: [.qr], completion: processQRCode)
                    .aspectRatio(1, contentMode: .fit)
                    .border(.gray)
            }
            Text("You can also add a server address manually.")
        }
```
(this is a bit redundant, maybe something else)

File: apps/ios/SimpleXChat/APITypes.swift [hunk 6]
```diff
@@ -701,24 +721,28 @@ public enum SMPTestStep: String, Decodable {
...
+    public static func == (l: SMPTestFailure, r: SMPTestFailure) -> Bool {
+        l.testStep == r.testStep
+    }
+
+    public var localizedDescription: String {
+        let err = String.localizedStringWithFormat(NSLocalizedString("Test failed at step %@.", comment: "server test failure"), testStep.text)
         switch testError {
         case .SMP(.AUTH):
-            return err + "," + NSLocalizedString("Server requires authentication to create queues, check password", comment: "server test error")
+            return err + NSLocalizedString(" Server requires authorization to create queues, check password", comment: "server test error")
```
Comment (jr-simplex) [inline apps/ios/SimpleXChat/APITypes.swift:736]: The space in the beginning is strange, and easy to miss in translations
```suggestion
            return err + " " + NSLocalizedString("Server requires authorization to create queues, check password", comment: "server test error")
```

File: apps/ios/SimpleXChat/APITypes.swift [hunk 7]
```diff
@@ -701,24 +721,28 @@ public enum SMPTestStep: String, Decodable {
...
         case .BROKER(.NETWORK):
-            return err + "," + NSLocalizedString("Possibly, certificate fingerprint in server address is incorrect", comment: "server test error")
+            return err + NSLocalizedString(" Possibly, certificate fingerprint in server address is incorrect", comment: "server test error")
```
Comment (jr-simplex) [inline apps/ios/SimpleXChat/APITypes.swift:738]: same
```suggestion
            return err + " " + NSLocalizedString("Possibly, certificate fingerprint in server address is incorrect", comment: "server test error")
```

File: apps/ios/Shared/Views/UserSettings/ScanSMPServer.swift (hunk 5 above)
Comment (epoberezkin) [reply apps/ios/Shared/Views/UserSettings/ScanSMPServer.swift:29]: That'll do for now :) I thought about it, but decided not to clutter it with the explanations - it's all pretty clear here.

File: apps/ios/SimpleX Localizations/ru.xcloc/Localized Contents/ru.xliff [hunk 8]
```diff
@@ -27,6 +27,16 @@
         <target> (можно скопировать)</target>
         <note>No comment provided by engineer.</note>
       </trans-unit>
+      <trans-unit id="Possibly, certificate fingerprint in server address is incorrect" xml:space="preserve">
+        <source>Possibly, certificate fingerprint in server address is incorrect</source>
+        <target> Возможно, хэш сертификата адресе не совпадает с сертификатом</target>
```
Comment (jr-simplex) [inline apps/ios/SimpleX Localizations/ru.xcloc/Localized Contents/ru.xliff:32]:
```suggestion
        <target>Возможно, хэш сертификата в адресе сервера неверный</target>
```

File: apps/ios/SimpleX Localizations/ru.xcloc/Localized Contents/ru.xliff [hunk 9]
```diff
@@ -27,6 +27,16 @@
...
+        <note>server test error</note>
+      </trans-unit>
+      <trans-unit id="Server requires authorization to create queues, check password" xml:space="preserve">
+        <source>Server requires authorization to create queues, check password</source>
+        <target> Сервер требует авторизации для создания очередей, проверьте пароль</target>
```
Comment (jr-simplex) [inline apps/ios/SimpleX Localizations/ru.xcloc/Localized Contents/ru.xliff:37]:
```suggestion
        <target>Сервер требует авторизации для создания очередей, проверьте пароль</target>
```

File: apps/ios/SimpleX Localizations/en.xcloc/Localized Contents/en.xliff [hunk 10]
```diff
@@ -531,7 +556,7 @@
       <trans-unit id="Connect" xml:space="preserve">
         <source>Connect</source>
         <target>Connect</target>
-        <note>No comment provided by engineer.</note>
+        <note>server test step</note>
```
Comment (jr-simplex) [inline apps/ios/SimpleX Localizations/en.xcloc/Localized Contents/en.xliff:559]: Maybe create a separate key? "Server test step Connect" and change English translation? At least we should check translations still work, I can do it later

File: apps/ios/SimpleX Localizations/ru.xcloc/Localized Contents/ru.xliff [hunk 11]
```diff
@@ -2448,6 +2523,26 @@ We will be adding server redundancy to prevent lost messages.</source>
...
+        <note>server test failure</note>
+      </trans-unit>
+      <trans-unit id="Test server" xml:space="preserve">
+        <source>Test server</source>
+        <target>Тест сервер</target>
+        <note>No comment provided by engineer.</note>
+      </trans-unit>
+      <trans-unit id="Test servers" xml:space="preserve">
+        <source>Test servers</source>
+        <target>Тест серверы</target>
```
Comment (jr-simplex) [inline apps/ios/SimpleX Localizations/ru.xcloc/Localized Contents/ru.xliff:2538]: or "Тест серверов"
```suggestion
        <target>Протестировать серверы</target>
```

**Pull request merged**